piet-wgsl/shader/fine.wgsl

```diff
@@ -281,7 +281,10 @@ fn main(
     for (var i = 0u; i < PIXELS_PER_THREAD; i += 1u) {
         let coords = xy_uint + vec2(i, 0u);
         if coords.x < config.target_width && coords.y < config.target_height {
-            textureStore(output, vec2<i32>(coords), rgba[i]);
+            let fg = rgba[i];
+            let a_inv = 1.0 / (fg.a + 1e-6);
+            let rgba_sep = vec4(fg.r * a_inv, fg.g * a_inv, fg.b * a_inv, fg.a);            
```

Comment:
This probably isn't a meaningful question, but why does an earlier step do the inverse of this multiplication?

It feels like it would be more meaningful to just avoid having that done too early into what is stored?

Reply:
Raph and I had a long conversation about this at some point and the upshot is that storing in premultiplied form is a few ALU ops shorter for basic source over and thus faster in the common case. We have to handle the conversion in blends on either input or output regardless, so this was chosen as the optimal strategy.

Reply:
It's not just one earlier step, it's an arbitrary amount of blending and compositing, which will be Porter-Duff "over" operations in the common case. It's both more efficient and more correct to do this operation premultiplied space. Best source for that is probably [Tom Forsyth's blog](https://tomforsyth1000.github.io/blog.wiki.html) (search for "premultiplied alpha, internal links seem to be broken). There are a few more links in my [2d.graphics page](https://2d.graphics/book/compositing.html), which I'd really like to flesh out some time.

As we've been discussing, it makes sense for this shader to produce either premultiplied or separated alpha depending on how its output will be used. For most use cases where it is consumed by the GPU, premultiplied alpha is better. For storing in a PNG, separated alpha is better. The *simplest* thing to do is to output separated here (basically at the cost of that divide instruction), then convert to premultiplied alpha downstream, and I'm ok with leaving it in that state.

The premultiplied alpha question also has a complex interaction with gamma. That's out of scope for now but is something we will want to address at some time.